core/embed/trezorhal/stm32f4/backlight_pwm.c

```diff
@@ -1,212 +1,280 @@
+    // Set the Output Idle state
     tmpcr2 |= TIM_OCIDLESTATE_SET;
-    /* Set the Output N Idle state */
+    // Set the Output N Idle state
     tmpcr2 |= TIM_OCNIDLESTATE_SET;
   }
 
-  /* Write to TIMx CR2 */
+  // Write to TIMx CR2
   BACKLIGHT_PWM_TIM->CR2 = tmpcr2;
-
-  /* Write to TIMx CCMR1 */
+  // Write to TIMx CCMR1
   BACKLIGHT_PWM_TIM->CCMR1 = tmpccmrx;
-
-  /* Set the Capture Compare Register value */
+  // Set the Capture Compare Register value
   BACKLIGHT_PWM_TIM->CCR1 = 0;
-
-  /* Write to TIMx CCER */
+  // Write to TIMx CCER
   BACKLIGHT_PWM_TIM->CCER = tmpccer;
 
-  backlight_pwm_set(0);
-
   BACKLIGHT_PWM_TIM->BDTR |= TIM_BDTR_MOE;
   BACKLIGHT_PWM_TIM->CR1 |= TIM_CR1_CEN;
+
+  drv->initialized = true;
+
+  backlight_pwm_set(initial_level);
 }
 
-void backlight_pwm_reinit(void) {
-  uint32_t prev_arr = BACKLIGHT_PWM_TIM->ARR;
-  uint32_t prev_ccr1 = BACKLIGHT_PWM_TIM->BACKLIGHT_PWM_TIM_CCR;
+void backlight_pwm_deinit(backlight_action_t action) {
+  backlight_driver_t *drv = &g_backlight_driver;
 
-  uint32_t prev_val = (prev_ccr1 * 255) / (prev_arr + 1);
-  prev_val = prev_val > 255 ? 255 : prev_val;
-  BACKLIGHT = prev_val;
+  if (!drv->initialized) {
+    return;
+  }
 
-  pwm_period = LED_PWM_TIM_PERIOD;
-  BACKLIGHT_PWM_TIM->PSC = LED_PWM_PRESCALER;
-  BACKLIGHT_PWM_TIM->CR1 |= TIM_CR1_ARPE;
-  BACKLIGHT_PWM_TIM->CR2 |= TIM_CR2_CCPC;
-  BACKLIGHT_PWM_TIM->BACKLIGHT_PWM_TIM_CCR = (pwm_period * prev_val) / 255;
-  BACKLIGHT_PWM_TIM->ARR = LED_PWM_TIM_PERIOD - 1;
-}
+  if (action == BACKLIGHT_RETAIN) {
+    // We keep both the GPIO and the timer running
 
 #ifdef TREZOR_MODEL_T
+// This code here is for backward compatibility with the old
+// bootloader that used a different PWM settings.
+
+    // about 10Hz (with PSC = (SystemCoreClock / 1000000) - 1)
+    #define LED_PWM_SLOW_TIM_PERIOD (10000)
+    #define LED_PWM_PRESCALER_SLOW (SystemCoreClock / 1000000 - 1)  // 1 MHz
+
+    BACKLIGHT_PWM_TIM->PSC = LED_PWM_PRESCALER_SLOW;
+    BACKLIGHT_PWM_TIM->CR1 |= TIM_CR1_ARPE;
+    BACKLIGHT_PWM_TIM->CR2 |= TIM_CR2_CCPC;
+    BACKLIGHT_PWM_TIM->ARR = LED_PWM_SLOW_TIM_PERIOD - 1;
+    BACKLIGHT_PWM_TIM->BACKLIGHT_PWM_TIM_CCR =
+        (LED_PWM_SLOW_TIM_PERIOD * drv->current_level) / 255;
+#endif
+  } else {
+    // TODO: deinitialize GPIOs and the TIMER
+  }
 
-#define LED_PWM_SLOW_TIM_PERIOD \
-  (10000)  // about 10Hz (with PSC = (SystemCoreClock / 1000000) - 1)
+  drv->initialized = false;
+}
 
-void backlight_pwm_set_slow(void) {
-  uint32_t prev_arr = BACKLIGHT_PWM_TIM->ARR;
-  uint32_t prev_ccr1 = BACKLIGHT_PWM_TIM->CCR1;
+// Generate a pulse on the backlight control pin to wake up the TPS61043
+static void backlight_pwm_wakeup_pulse(void) {
+  GPIO_InitTypeDef GPIO_InitStructure = {0};
 
-  uint32_t prev_val = (prev_ccr1 * 255) / (prev_arr + 1);
-  prev_val = prev_val > 255 ? 255 : prev_val;
+  HAL_GPIO_WritePin(BACKLIGHT_PWM_PORT, BACKLIGHT_PWM_PIN, GPIO_PIN_SET);
+  // LCD_PWM/PA7 (backlight control)
```

Comment:
the PA7 note is outdated as this can be on different pins too (i know it not introduced here but lets fix it while we are touching this)

Reply:
fixed, see 5632770f21a629eca8115321d811da2cc66f9538